File: user-service/src/main/java/rs/edu/raf/service/impl/EmployeeServiceImpl.java
package rs.edu.raf.service.impl;

import io.micrometer.common.util.StringUtils;
import lombok.AllArgsConstructor;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import rs.edu.raf.annotations.GeneratedCrudOperation;
import rs.edu.raf.dto.EditEmployeeDTO;
import rs.edu.raf.dto.EmployeeCreateDTO;
import rs.edu.raf.dto.EmployeeDTO;
import rs.edu.raf.exceptions.UserNotFoundException;
import rs.edu.raf.mapper.EmployeeMapper;
import rs.edu.raf.model.user.Employee;
import rs.edu.raf.repository.EmployeeRepository;
import rs.edu.raf.repository.RoleRepository;
import rs.edu.raf.service.EmployeeService;
import rs.edu.raf.util.JMBGValidator;

import java.util.List;
import java.util.function.Consumer;

@Service
@AllArgsConstructor
public class EmployeeServiceImpl implements EmployeeService {
    private EmployeeRepository employeeRepository;
    private EmployeeMapper employeeMapper;
    private BCryptPasswordEncoder bCryptPasswordEncoder;
    private RoleRepository roleRepository;

    @Override
    public EmployeeDTO createNewEmployee(EmployeeCreateDTO employeeCreateDTO) {

        Employee employee = employeeMapper.employeeCreateDTOtoEmployee(employeeCreateDTO);

        JMBGValidator.validateJMBG(employee.getDateOfBirth(), employee.getJMBG());

        return employeeMapper.employeeToEmployeeDTO(employeeRepository.save(employee));
    }

    @Override
    public EmployeeDTO editEmployee(EditEmployeeDTO editEmployeeDTO) {

        Employee employee = employeeRepository.findById(editEmployeeDTO.id())
                .orElseThrow(() -> new UserNotFoundException("Employee with id " + editEmployeeDTO.id() + " not found!"));

        updateIfPresent(employee::setLastName, editEmployeeDTO.lastName());
        updateIfPresent(employee::setGender, editEmployeeDTO.gender());
        updateIfPresent(employee::setPhoneNumber, editEmployeeDTO.phoneNumber());
        updateIfPresent(employee::setPassword, bCryptPasswordEncoder.encode(editEmployeeDTO.password()));
        updateIfPresent(employee::setPosition, editEmployeeDTO.position());
        updateIfPresent(employee::setAddress, editEmployeeDTO.address());
        updateIfPresent(employee::setDepartment, editEmployeeDTO.department());
        employee.setActive(editEmployeeDTO.active());

        if(editEmployeeDTO.roles() != null) {
            employee.setRoles(editEmployeeDTO.roles().stream().map(role -> roleRepository.findByName(role)
                    .orElseThrow()).toList());
        }

        return employeeMapper.employeeToEmployeeDTO(employeeRepository.save(employee));
    }

    @Override
    public List<EmployeeDTO> getAllActiveEmployees() {
        return employeeRepository.findAllByActiveIsTrue().stream().map(employeeMapper::employeeToEmployeeDTO).toList();
    }

    @Override
    public EmployeeDTO findActiveEmployeeByEmail(String email) {
        return employeeMapper.employeeToEmployeeDTO(employeeRepository
                .findByEmailAndActiveIsTrue(email)
                .orElseThrow(()->new UserNotFoundException("Employee with email " + email + " not found!")));
    }

    @GeneratedCrudOperation
    @Override
    public EmployeeDTO findActiveEmployeeById(Long id){
        return employeeMapper.employeeToEmployeeDTO(employeeRepository
                .findById(id)
                .orElseThrow(() -> new UserNotFoundException("Employee with ID " + id + " not found!")));
    }

    private void updateIfPresent(Consumer<String> setter, String value) {
        if(StringUtils.isNotBlank(value))
            setter.accept(value);
    }
}

File: user-service-kotlin/src/main/kotlin/org/example/userservicekotlin/service/impl/EmployeeServiceImpl.kt
package org.example.userservicekotlin.service.impl

import org.example.userservicekotlin.dto.EditEmployeeDTO
import org.example.userservicekotlin.dto.EmployeeCreateDTO
import org.example.userservicekotlin.dto.EmployeeDTO
import org.example.userservicekotlin.exception.UserNotFoundException
import org.example.userservicekotlin.mapper.EmployeeMapper
import org.example.userservicekotlin.model.user.Employee
import org.example.userservicekotlin.repository.EmployeeRepository
import org.example.userservicekotlin.repository.RoleRepository
import org.example.userservicekotlin.service.EmployeeService
import org.example.userservicekotlin.util.JMBGValidator
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service

@Service
class EmployeeServiceImpl(
    val employeeRepository: EmployeeRepository,
    val employeeMapper: EmployeeMapper,
    val roleRepository: RoleRepository,
    val bCryptPasswordEncoder: BCryptPasswordEncoder
) : EmployeeService {
    override fun createNewEmployee(employeeCreateDTO: EmployeeCreateDTO): EmployeeDTO {
        val employee: Employee = employeeMapper.employeeCreateDTOtoEmployee(employeeCreateDTO)

        JMBGValidator.validateJMBG(employee.dateOfBirth, employee.JMBG)

        return employeeMapper.employeeToEmployeeDTO(employeeRepository.save(employee))
    }

    override fun editEmployee(editEmployeeDTO: EditEmployeeDTO): EmployeeDTO {
        val employee: Employee = employeeRepository.findById(editEmployeeDTO.id)
            .orElseThrow { UserNotFoundException("Employee with id " + editEmployeeDTO.id + " not found!") }

        updateIfPresent({ employee.lastName = it }, editEmployeeDTO.lastName)
        updateIfPresent({ employee.gender = it }, editEmployeeDTO.gender)
        updateIfPresent({ employee.phoneNumber = it }, editEmployeeDTO.phoneNumber)
        updateIfPresent({ employee.userPassword = it }, bCryptPasswordEncoder.encode(editEmployeeDTO.password))
        updateIfPresent({ employee.position = it }, editEmployeeDTO.position)
        updateIfPresent({ employee.address = it }, editEmployeeDTO.address)
        updateIfPresent({ employee.department = it }, editEmployeeDTO.department)
        employee.active = editEmployeeDTO.active

        employee.roles = editEmployeeDTO.roles.map { role ->
            roleRepository.findByName(role)
                .orElseThrow()
        }.toMutableList()

        return employeeMapper.employeeToEmployeeDTO(employeeRepository.save(employee))
    }

    override fun getAllActiveEmployees(): List<EmployeeDTO> {
        return employeeRepository.findAllByActiveIsTrue().stream().map(employeeMapper::employeeToEmployeeDTO).toList()
    }

    override fun findActiveEmployeeByEmail(email: String): EmployeeDTO {
        return employeeMapper.employeeToEmployeeDTO(
            employeeRepository
                .findByEmailAndActiveIsTrue(email)
                .orElseThrow { UserNotFoundException("Employee with email $email not found!") })
    }

    override fun findActiveEmployeeById(id: Long): EmployeeDTO {
        return employeeMapper.employeeToEmployeeDTO(
            employeeRepository
                .findById(id)
                .orElseThrow { UserNotFoundException("Employee with ID $id not found!") })
    }

    fun updateIfPresent(setter: (String) -> Unit, value: String?) {
        if (!value.isNullOrBlank()) setter(value)
    }
}